### ISSUE TITLE
Errors returned are not typed making it difficult to distinguish for approriate handling.

### ISSUE THREAD

**Issue by sjhx** [containers/image#1110]
Specifically `ImageReferece.NewImageSource()` error can indicate that the credentials are rejected (permanent) or alternatively that there is a transient network error e.g. "dial connection timeout", "tls timeout" or 429 "too many requests" (transient).

Without string matching it is not possible to distinguish one from another and in order to handle approriately. 

Ref: https://github.com/IBM/portieris/issues/238

**Comment by mtrmac** [comment]
Specifically for the Docker client, there is at least https://github.com/containers/image/blob/122b16af75d13b9dacda1b04111fa7fdffde30c5/docker/errors.go#L12-L23 . That’s not a general solution, but probably enough at least for the “credentials rejected, give up” case.

https://github.com/containers/common/tree/master/pkg/retry might be useful as well.

**Comment by jhart1685** [comment]
Another example of an error (from `image/docker/errors.go`) that should be typed is `manifest unknown`.

When writing an application (using this library), I want to be able to ignore cases where source image does not exist. Currently the only way to do this is with a string compare for `manifest unknown`.
It would be extremely helpful if the containers image library returned a specific error (or typed error) in this case.

**Comment by mtrmac** [comment]
That’s a reasonable request.

`manifest unknown` is actually, _currently_, typed as `github.com/docker/distribution/registry/api/v2.ErrorCodeManifestUnknown` — but that’s not an API commitment, and it’s not something we’d want to _make_ an API commitment, because registries behave inconsistently in this (IIRC some just report a HTTP 404 status). So ideally c/image should hide these registry differences and provide a new error type (i.e. _not_ the `api/v2` one above).

But then there are semantic difficulties — some registries (in some configurations?) only ever report “unauthorized” / “denied”, never “manifest unknown”, because users who are not allowed to access images are not allowed to know whether that image exists either. So it’s unclear what the library can do about that.

Finally, this would require some kind of interoperability testing infrastructure (which we should have anyway, but sadly we don’t) to be able to make a reasonable API commitment of any kind.

**Comment by jhart1685** [comment]
Hi @mtrmac - thanks for the response, and apologies for the delay in getting back to you.
You make good points about differences between different registry implementations, the need for a new error type, and the requirement for interoperability testing.

I'm not sure I understand your first point though:
> manifest unknown is actually, currently, typed as github.com/docker/distribution/registry/api/v2.ErrorCodeManifestUnknown — but that’s not an API commitment

Are you suggesting that (in some cases at least), we should get a typed error back from containers/image?

I've just tested with the latest docker distribution code, by running a plain, insecure registry locally (and pushed some images into it):
```
docker run -d -p 5001:5000 --restart always --name registry registry:latest
```

Here's a code snippet that will use `copy.Image` to copy an image from one reference in that registry to another:
https://gist.github.com/jhart1685/58af75f97a1022ae188adc49cab8d0a8
The code references a src manifest that does not exist (I checked that the code _works_ with an existing src image).
In the error handling code, I iterate around an `Unwrap`, writing the error and error type at each point in the stack until it can no longer be unwrapped.

The output looks like:
```
*errors.withStack, initializing source docker://localhost:5001/idontexist:1: reading manifest 1 in localhost:5001/idontexist: manifest unknown: manifest unknown
*errors.withMessage, initializing source docker://localhost:5001/idontexist:1: reading manifest 1 in localhost:5001/idontexist: manifest unknown: manifest unknown
*errors.withStack, reading manifest 1 in localhost:5001/idontexist: manifest unknown: manifest unknown
*errors.withMessage, reading manifest 1 in localhost:5001/idontexist: manifest unknown: manifest unknown
errcode.Errors, manifest unknown: manifest unknown
```

No sign of a `v2.ErrorCodeManifestUnknown` in that stack.

**Comment by mtrmac** [comment]
@jhart1685 Yes, I’m basically suggesting that, except that `ErrorCodeManifestUnknown` is not quite an `error`. Compare the recently-added https://github.com/containers/image/blob/29aec5f7bbbf6d58cf377f7d2aaf9ac32ba2bfa0/docker/docker_client.go#L923-L934. Caveats:
- It’s very likely that _doesn’t work_ reliably against the full universe of registries.
- c/image _does not_ document or promise that it will continue returning that return type; compare https://github.com/containers/image/pull/1487 which already proposes to change some parts of the returned error value.

**Comment by jhart1685** [comment]
@mtrmac - ok, thanks. Based on results I'm seeing, I don't believe that new `isManifestUnknownError` is in the code path for the cases I've been testing. 
However, as you say, even if it were, we shouldn't be relying on it.
